### ISSUE TITLE
customArrows props creates infinite loops.

### ISSUE THREAD

**Issue by LalaSimon** [Clariity/react-chessboard#94]
Hi! 

Im trying to create chessboard site that allows users analyse positions. Main functionality is to let users draw arrows on chessboard, also to other users (using socket.io for moves, also trying it for arrows). Unfortunately when im trying to get data from onArrowsChange and broadcast it to the other users, while trying to set customArrows from useState i got infinite loop and site is getting crashed. It seems like a bug for me but maybe its just too tricky for me. How should i implement customArrows props from server without getting crushed?

**Comment by Manukyanq** [comment]
Hi! 

take a look to [useArrows.ts](https://github.com/Clariity/react-chessboard/blob/main/src/chessboard/hooks/useArrows.ts) which is used for arrows state managment. 
if you change `customArrows` inside your `onArrowsChange` function, it might be the cause of infinitive loop.

Sharing your code will be much helpful for understanding whats exactly is going on

**Comment by LalaSimon** [comment]
Thank you for fast answer, i will try my best, so onArrowsChange im creating this function arrowDrow:

```typescript
const [arrows, setArrows] = useState<Square[][]>([]);
const arrowDrow = (arrowsData: Square[][]) => {
        if (arrowsData.length === 0 && arrowsData !== arrows) {
            setArrows([]);
            socket.emit("draw_arrows", {
                arrowsData,
                roomID,
            });
        } else {
            setArrows(arrowsData);
            socket.emit("draw_arrows", {
                arrowsData,
                roomID,
            });
        }
    };
``` 
My server is responding with broadcast like this:
```javascript
 socket.on("draw_arrows", (data) => {
        socket.broadcast.to(data.roomID).emit("arrows_drawn", data.arrowsData);
    });
``` 
 And then client received it with this function: 
```typescript
socket.on("arrows_drawn", (arrowsData) => {
        setArrows(arrowsData);
    });
``` 

My component is rendered like this:
```typescript
 <Chessboard
                        customArrows={arrows}
                        onArrowsChange={arrowDrow}
                        onPieceDrop={onDrop}
                        position={game.fen()}
                        boardWidth={650}
                    />
                    ```

**Comment by Manukyanq** [comment]
Yes, my guess seems right, when you call `onArrowsChange` it sets new state by triggering `setArrows(arrowsData);`. after that here in [useArrows.ts](https://github.com/Clariity/react-chessboard/blob/main/src/chessboard/hooks/useArrows.ts#L31)  we trigger `onArrowsChange` again it causes infinite loop. 

I would think about correct solution,  have to discuss with @Clariity what we can do

As an dirty temporary hack  i will suggest you to add extra check in your function

```
const arrowDrow = (arrowsData: Square[][]) => {
        if (arrowsData.length === 0 && arrowsData !== arrows) {
            setArrows([]);
            socket.emit("draw_arrows", {
                arrowsData,
                roomID,
            });
        } else {
            // this should stop endless loop
            if(arrowsData.flat().join() === arrows.flat().join()) return

            setArrows(arrowsData);
            socket.emit("draw_arrows", {
                arrowsData,
                roomID,
            });
        }
    };
```

**Comment by LalaSimon** [comment]
Thank you very much @Manukyanq, it works now!!

**Comment by LalaSimon** [comment]
I tried it but it seems not working. First arrow is good but next ones are not showing for both users
```typescript
const arrowDrow = (arrowsData: Square[][]) => {
            socket.emit("draw_arrows", {
                arrowsData,
                roomID,
            });
    };
```
I think i need to call setArrorws because i broadcast it on server side so its not coming back. When i tried to emit it instead of broadcasting its not working.

**Comment by Clariity** [comment]
Thank you for responding whilst I was away on holiday @Manukyanq 

For the line you references in useArrows.ts:

```
// callback when new arrows are set
  useEffect(() => {
    onArrowsChange?.(toArray(arrows));
  }, [arrows]);
```

I think we need to change the implementation so that this is only called when the change comes from within the board, and not from an external arrows prop change, and the name should be changed as to not cause confusion i.e. onArrowsBoardChange

this is because if it's coming externally from then the user can handle the change externally

**Comment by Clariity** [comment]
fixed in #99 

will release soon
